src/nbl/asset/utils/CHLSLCompiler.cpp

```diff
@@ -25,22 +25,185 @@
+        public:
+            template <typename PositionT>
+            static void init_iterators(IterContextT& iter_ctx,
+                PositionT const& act_pos, boost::wave::language_support language)
+            {
+                typedef typename IterContextT::iterator_type iterator_type;
+
+                std::string s(iter_ctx.filename.begin(), iter_ctx.filename.end());
+                auto icf = iter_ctx.ctx.get_hooks().preprocessOptions.includeFinder; // TODO it's null?
+                //std::optional<std::string> st = icf.getIncludeStandard(p, s);
+                //unfortunately have to call get_resource_runtime here again
+                auto builtin_pair = nbl::builtin::get_resource_runtime(s);
+                if (builtin_pair.first) //not a nullptr
+                {
+                    iter_ctx.instring = std::string(builtin_pair.first, builtin_pair.first+builtin_pair.second);
+                }
+                else {
+                    //TODO use nbl filesystem
+                    // read in the file
+                    boost::filesystem::ifstream instream(iter_ctx.filename.c_str());
+                    if (!instream.is_open()) {
+                        BOOST_WAVE_THROW_CTX(iter_ctx.ctx, boost::wave::preprocess_exception,
+                            bad_include_file, iter_ctx.filename.c_str(), act_pos);
+                        return;
+                    }
+                    instream.unsetf(std::ios::skipws);
+
+                    iter_ctx.instring.assign(
+                        std::istreambuf_iterator<char>(instream.rdbuf()),
+                        std::istreambuf_iterator<char>());
+                }
+                iter_ctx.first = iterator_type(
+                    iter_ctx.instring.begin(), iter_ctx.instring.end(),
+                    PositionT(iter_ctx.filename), language);
+                iter_ctx.last = iterator_type();
+            }
+
+        private:
+            std::string instring; //todo change to IReferenceCounted file
+        };
+    };
+
+    struct custom_preprocessing_hooks : public boost::wave::context_policies::default_preprocessing_hooks{
+
+        custom_preprocessing_hooks(const nbl::asset::IShaderCompiler::SPreprocessorOptions& _preprocessOptions, std::vector<std::pair<uint32_t, std::string>>&& _lineOffsetStack) : preprocessOptions(_preprocessOptions), lineOffsetStack(_lineOffsetStack) {}
+
+        nbl::asset::IShaderCompiler::SPreprocessorOptions preprocessOptions;
+        std::vector<std::pair<uint32_t, std::string>> lineOffsetStack;  //TODO no longer needed
+
+
+        template <typename ContextT>
+        bool
+            locate_include_file(ContextT& ctx, std::string& file_path,
+                bool is_system, char const* current_name, std::string& dir_path,
+                std::string& native_name) {
+
+            //on builtin return true
+            //default returns false if file does not exist
+            if (nbl::builtin::hasPathPrefix(file_path))
```

Comment:
there's many multiple types of builtins copy the old code verbatim
```cpp
    std::filesystem::path relDir;
    #ifdef NBL_EMBED_BUILTIN_RESOURCES
    const bool reqFromBuiltin = nbl::builtin::hasPathPrefix(requestingSource) || spirv::builtin::hasPathPrefix(requestingSource);
    const bool reqBuiltin = nbl::builtin::hasPathPrefix(requestedSource) || spirv::builtin::hasPathPrefix(requestedSource);
    if (!reqFromBuiltin && !reqBuiltin)
    {
        //While #includ'ing a builtin, one must specify its full path (starting with "nbl/builtin" or "/nbl/builtin").
        //  This rule applies also while a builtin is #includ`ing another builtin.
        //While including a filesystem file it must be either absolute path (or relative to any search dir added to asset::iIncludeHandler; <>-type),
        //  or path relative to executable's working directory (""-type).
        relDir = std::filesystem::path(requestingSource).parent_path();
    }
    #else
    const bool reqBuiltin = false;
    #endif // NBL_EMBED_BUILTIN_RESOURCES
    std::filesystem::path name = isRelative ? (relDir / requestedSource) : (requestedSource);

    if (std::filesystem::exists(name) && !reqBuiltin)
        name = std::filesystem::absolute(name);
```

although you should probably replace `std::filesystem::path` with `nbl::system::path`

Reply:
you can use the old `name` to use as `file_path` return value